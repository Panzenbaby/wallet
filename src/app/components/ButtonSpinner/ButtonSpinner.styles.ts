import { FlattenSimpleInterpolation } from "styled-components";
import tw, { css, TwStyle } from "twin.macro";
import { ButtonVariant } from "types";
import { shouldUseDarkColors } from "utils/electron-utils";

const baseStyle = [tw`w-6 h-6 border border-4 rounded-full animate-spin`];

const variantColors: {
	[key in ButtonVariant]?: string;
} = {
	danger: "danger-400",
	primary: "white",
	secondary: "primary-600",
};

const getColor = (variant: ButtonVariant): Array<FlattenSimpleInterpolation | TwStyle> => {
	const color = variantColors[variant] || "primary-500";

	const styles: Array<FlattenSimpleInterpolation | TwStyle> = [];

	switch (variant) {
		case "danger": {
			styles.push(tw`border-theme-danger-200 dark:border-theme-danger-500`);

			break;
		}
		case "primary": {
			styles.push(tw`border-theme-primary-700`);

			break;
		}
		case "secondary": {
			styles.push(tw`border-white dark:border-theme-secondary-900`);

			break;
		}
		default: {
			styles.push(tw`border-theme-secondary-200`);
		}
	}

	if (variant === "danger" && shouldUseDarkColors()) {
		styles.push(
			css`
				border-left-color: var(--theme-color-white) !important;
			`,
		);
	} else {
		styles.push(
			css`
				border-left-color: var(--theme-color-${color}) !important;
			`,
		);
	}

	return styles;
};

export const getStyles = ({ variant }: { variant?: ButtonVariant }) => [...baseStyle, ...getColor(variant!)];
