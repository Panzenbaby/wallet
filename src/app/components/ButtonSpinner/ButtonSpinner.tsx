import { styled } from "twin.macro";
import { ButtonVariant } from "types";

import { getStyles } from "./ButtonSpinner.styles";

interface ButtonSpinner {
	variant: ButtonVariant | undefined;
}

export const ButtonSpinner = styled.div<ButtonSpinner>(getStyles);

ButtonSpinner.defaultProps = {
	variant: "primary",
};
