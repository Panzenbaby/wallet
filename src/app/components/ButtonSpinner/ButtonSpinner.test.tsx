import React from "react";

import { ButtonSpinner } from "./ButtonSpinner";
import { render } from "@/utils/testing-library";

describe("ButtonSpinner", () => {
	beforeAll(() => {
		jest.spyOn(console, "error").mockImplementation(() => null);
	});

	afterAll(() => {
		console.error.mockRestore();
	});

	it("should render", () => {
		const { container } = render(<ButtonSpinner />);

		expect(container).toMatchSnapshot();
	});

	it.each(["primary", "secondary", "danger", "transparent", "info", "reverse"])(
		"should render a %s variant",
		(variant) => {
			const { container } = render(<ButtonSpinner color={variant} />);

			expect(container).toMatchSnapshot();
		},
	);
});
