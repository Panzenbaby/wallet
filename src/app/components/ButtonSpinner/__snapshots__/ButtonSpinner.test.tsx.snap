// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonSpinner should render 1`] = `
<div>
  <div
    class="ButtonSpinner-sc-1jf6tdt-0 jgnpkQ"
  />
</div>
`;

exports[`ButtonSpinner should render a danger variant 1`] = `
<div>
  <div
    class="ButtonSpinner-sc-1jf6tdt-0 jgnpkQ"
    color="danger"
  />
</div>
`;

exports[`ButtonSpinner should render a info variant 1`] = `
<div>
  <div
    class="ButtonSpinner-sc-1jf6tdt-0 jgnpkQ"
    color="info"
  />
</div>
`;

exports[`ButtonSpinner should render a primary variant 1`] = `
<div>
  <div
    class="ButtonSpinner-sc-1jf6tdt-0 jgnpkQ"
    color="primary"
  />
</div>
`;

exports[`ButtonSpinner should render a reverse variant 1`] = `
<div>
  <div
    class="ButtonSpinner-sc-1jf6tdt-0 jgnpkQ"
    color="reverse"
  />
</div>
`;

exports[`ButtonSpinner should render a secondary variant 1`] = `
<div>
  <div
    class="ButtonSpinner-sc-1jf6tdt-0 jgnpkQ"
    color="secondary"
  />
</div>
`;

exports[`ButtonSpinner should render a transparent variant 1`] = `
<div>
  <div
    class="ButtonSpinner-sc-1jf6tdt-0 jgnpkQ"
    color="transparent"
  />
</div>
`;
